### PR TITLE
Ensure ResultMessage is processed

### PR DESCRIPTION
**Pull request by jiangpengcheng**
## Description
As `processCompeltion` and `processResult` is executed asynchronously, it's better to ensure `ResultMessage` will be processed in any case

## Related issue and scope
- [x] I opened an issue to propose and discuss this change (#4134 )

## My changes affect the following components
- [ ] API
- [ ] Controller
- [ ] Message Bus (e.g., Kafka)
- [x] Loadbalancer
- [ ] Invoker
- [ ] Intrinsic actions (e.g., sequences, conductors)
- [ ] Data stores (e.g., CouchDB)
- [ ] Tests
- [ ] Deployment
- [ ] CLI
- [ ] General tooling
- [ ] Documentation

## Types of changes
- [ ] Bug fix (generally a non-breaking change which closes an issue).
- [x] Enhancement or new feature (adds new functionality).
- [ ] Breaking change (a bug fix or enhancement which changes existing behavior).

## Checklist:

- [x] I signed an [Apache CLA](https://github.com/apache/incubator-openwhisk/blob/master/CONTRIBUTING.md).
- [x] I reviewed the [style guides](https://github.com/apache/incubator-openwhisk/wiki/Contributing:-Git-guidelines#code-readiness) and followed the recommendations (Travis CI will check :).
- [ ] I added tests to cover my changes.
- [ ] My changes require further changes to the documentation.
- [ ] I updated the documentation where necessary.

### CODE REVIEW

**Comment by cbickel** [comment]
@jiangpengcheng Thanks a lot for tackling this issue again.
I also looked at the problem again. What do you think about the following approach of fixing this:
The promise is only used in case the activation was blocking. In case of non-blocking activations, this promise is never used.
What do you think about only closing the promise on getting the ResultMessage?

**Comment by jiangpengcheng** [comment]
@cbickel 
do you mean create an extra `Promise[Either[ActivationId, WhiskActivation]]` , saved it into the new map for blocking activations, and return this `promise.future` in the `publish` method,
and for non-blocking activations, just keep the original way?

or

don't create promise for non-blocking activations, just return an `Future.successful(Left(aid))` in the `publish` method?

**Comment by cbickel** [comment]
Ah wait. One more thing came into my mind. Do we still need to store the promise in the `ActivationEntry` with this change? I think it would be enough to store it in `blockingPromises`, right?

**Comment by jiangpengcheng** [comment]
yes, we don't need that promise

**Comment by ningyougang** [comment]
@cbickel ,can be merged?

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 1]
```diff
@@ -422,13 +425,9 @@ class ShardingContainerPoolBalancer(
           .foreach(_.releaseConcurrent(entry.fullyQualifiedEntityName, entry.maxConcurrent, entry.memory.toMB.toInt))
         if (!forced) {
           entry.timeoutHandler.cancel()
```
Comment (rabbah) [inline core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: the timeout handler (which is the db poller) should be cancelled on the receipt of the active ack.

File: core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala [hunk 2]
```diff
@@ -262,9 +264,13 @@ class ShardingContainerPoolBalancer(
 
     chosen
       .map { invoker =>
-        val entry = setupActivation(msg, action, invoker)
+        setupActivation(msg, action, invoker)
         sendActivationToInvoker(messageProducer, msg, invoker).map { _ =>
-          entry.promise.future
+          if (msg.blocking) {
+            blockingPromises.getOrElseUpdate(msg.activationId, Promise[Either[ActivationId, WhiskActivation]]()).future
```
Comment (rabbah) [inline core/controller/src/main/scala/org/apache/openwhisk/core/loadBalancer/ShardingContainerPoolBalancer.scala]: while not likely, it's plausible the active ack is received before the promise is added to this map (which is why setup activation adds the promise before it's posted to any invoker).

**Pull request merged**